CRM/Core/Payment.php

```diff
@@ -1228,6 +1250,45 @@ public function doPayment(&$params, $component = 'contribute') {
     return $result;
   }
 
+  /**
+   * Refunds payment - this function wraps around doRefundPayment but first it checks wether the calling payment processor
+   *  supports refund with the help of supportsRefund()
+   *
+   * The function ensures an exception is thrown & moves some of this logic out of the form layer and makes the forms
+   * more agnostic.
+   *
+   * Payment processors should set payment_status_id if it set the status to Refunded in case the transaction is successful
+   *
+   * @param array $params
+   *
+   * @param string $component
+   *
+   * @return array
+   *   Result array
+   *
+   * @throws \Civi\Payment\Exception\PaymentProcessorException
+   */
+  public function doRefund(&$params, $component = 'contribute') {
+    $this->_component = $component;
```

Comment:
I don't think we need anything other than an empty function & a comment block here - it's up to the payment processors to implement this

The calling code won't call this without checking if the processor supports refund so we don't need to handle that here.

We don't have to grandfather this in so we don't have to have all the cruft in do payment - we could simply make the signature "anything other than a payment processor exception denotes success' (that will stop people being tempted to do weird stuff.)

Reply:
(the dummy class would have throw new PaymentProcessorException('processor doesn't support refunds)

Reply:
The concept of $component is a real pain and has lead to lot's of inconsistency/wierd handling in the past too.  Can we get rid of that parameter too and make it behave the same - currently we get different parameters available when it comes from event page, contribution page, API, webform etc. and it's "allowed" because you can test the component first.  

> "anything other than a payment processor exception denotes success'

Yes! The passing back of payment_status_id has led to a lot of confusion about who actually saves the result and whether you can pass other parameters too (hint for doPayment() via event you can pass other parameters, while when called by contribute you can't).

Also, can we have a list of mandatory/expected `$params` somewhere?

Agree the function here should be empty other than perhaps some commented best-practise code if there is anything that most processors are likely to do (for example, should the processor update the recur/contribution records themselves?

Reply:
"Also, can we have a list of mandatory/expected $params somewhere?"

yes we should have that - but apart from amount I'm not sure as yet that what is required for one processor will be the same as another at the moment

Reply:
I was actually thinking that amount wouldn't be a mandatory parameter (optional, but if specified in a standard format).  As you might just tell the processor to do a "full refund" of that transaction at which point the amount actually doesn't make sense, there could also be situations where the processor is actually refunding the client in a different currency so the amounts might come back differently.

Reply:
hmm - I feel like we need more examples to work through to get to that point.

Reply:
@monishdeb so just coming back to this - let's remove the ref to  $this->_component = $component; from input to this & being set - we would dearly love to remove all over the place but we can't but we can at least not add more